adding crates/README.md for rust docs (#4403)
roc-lang/roc: crates/README.md

> 

Reviewer: nit: remove leading newline

# Roc Internals

> Roc has different rust crates for various binaries and libraries. These are described below to help give you an overview. If you see something here that is out of date, please correct it or discuss it in the [Zulip chat](https://roc.zulipchat.com/).

Reviewer: Would style as something like

```suggestion
Roc has different rust crates for various binaries and libraries. Their roles are briefly described below. If you'd like to learn more, have any questions, or suspect something is out of date, please start a discussion on the [Roc Zulip](https://roc.zulipchat.com/)!
```

> Use `cargo doc` to generate docs; e.g. ```cargo doc --package roc_ast --open```.

Reviewer: ````suggestion
You can use `cargo doc` to generate docs for a specific package; e.g.

```
cargo doc --package roc_ast --open
```
````

## `ast/` - `roc_ast`

Code to represent the [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) as used by the editor.
In contrast to the compiler, the types in this AST do not keep track of the location of the matching code in the source file.

## `cli/` - `roc_cli`

The `roc` binary that brings together all functionality in the Roc toolset.

## `cli_utils/` - `cli_utils`

Provides shared code for cli tests and benchmarks.

## `code_markup/` - `roc_code_markup`

> Roc's very own [markup language](https://en.wikipedia.org/wiki/Markup_language). This library is used by the editor to display good-looking roc code. 

Reviewer: Would style as something like

```suggestion
A [markup language](https://en.wikipedia.org/wiki/Markup_language) to display Roc code in the editor
```

> ## `compile/`

Reviewer: ```suggestion
## `compiler/`
```

Compiles `.roc` files and combines them with their platform into an executable binary. See [compiler/README.md](./compiler/README.md) for more information.

TODO explain what "compiler frontend" is

> TODO explain what "compiler backend" is

Reviewer: Maybe worth skipping for now, this can be elaborated in a follow-up. Generally the compiler is the largest piece of the Roc project, a good introduction would be useful here; there is one we'd probably want to combine with this document in compiler/README.md. But, that doesn't need to be done in this patch.

The compiler includes the following sub-crates;

> - `roc_alias_analysis` TODO - Need assistance

Reviewer: Performs analyses and optimizations to remove unneeded reference counts at runtime, and support in-place mutation.

> - `arena-pool` TODO - Need assistance

Reviewer: An implementation of an arena allocator designed for the compiler's workloads.

> - `roc_build` TODO - Need assistance

Reviewer: Responsible for coordinating building and linking of a Roc app with its host.

- `roc_builtins` provdes the Roc functions and modules that are implicitly imported into every module. See [README.md](./compiler/builtins/README.md) for more information.

> - `roc_can` TODO - Need assistance

Reviewer: Canonicalizes a roc abstract syntax tree, resolving symbols, re-ordering definitions, and preparing a module for type inference.

> - `roc_collections` TODO - Need assistance

Reviewer: Domain-specific collections created for needs of the compiler.

> - `roc_constrain` TODO - Need assistance

Reviewer: Responsible for building the set of constraints that are used during type inference of a program, and for gathering context needed for pleasant error messages when a type error occurs.

> - `roc_debug_flags` TODO - Need assistance

Reviewer: Environment variables that can be toggled to aid debugging of the compiler itself.

> - `roc_derive` provides auto-derivers?

Reviewer: :+1:
```suggestion
- `roc_derive` provides auto-derivers for builtin abilities like `Hash` and `Decode`
```

- `roc_exhaustive` provides exhaustiveness checking for Roc 

> - `roc_fmt` TODO - Need assistance

Reviewer: The roc code formatter

- `roc_gen_dev` provides the compiler backend to generate Roc binaries fast, for a nice developer experience. See [README.md](./compiler/gen_dev/README.md) for more information.
- `roc_gen_llvm` provides the LLVM backend to generate Roc binaries. Used to generate a binary with the fastest possible execution speed.
- `roc_gen_wasm` provides the WASM backend to generate Roc binaries. See [README.md](./compiler/gen_wasm/README.md) for more information.

> - `roc_ident` provides identifiers

Reviewer: Implements data structures used for efficiently representing small strings, like identifiers.

> - `roc_exhaustive` provides exhaustiveness checking for Roc 

Reviewer: ```suggestion
- `roc_exhaustive` provides exhaustiveness checking for Roc.
```

- `roc_intern` provides generic interners for concurrent and single-thread use cases.
- `roc_late_solve` provides type unification and solving primitives from the perspective of the compiler backend.

> - `roc_load` used to load a `.roc` file and typecheck it.

Reviewer: Used to load a `.roc` file and coordinate the compiler pipeline, including parsing, type checking, and code generation.

> - `roc_load_internal` TODO - Need assistance

Reviewer: The internal implementation of `roc_load`, disjoint from `roc_load` to support caching.

> - `roc_module` TODO - Need assistance

Reviewer: Implements data structures used for efficiently representing unique modules and identifiers in Roc programs.

> - `roc_mono` TODO - Need assistance

Reviewer: Roc's main intermediate representation, which is responsible for monomorphization, defunctionalization, inserting ref-count instructions, and transforming a Roc program into a form that is easy to consume by a backend.

> - `roc_parse` TODO - Need assistance

Reviewer: Implements the Roc parser, which transforms a textual representation of a Roc program to an abstract syntax tree.

- `roc_problem` provides types to describe problems that can occur when compiling `.roc` code.
- `roc_region` TODO - Need assistance
- `roc_target` provides types and helpers for compiler targets such as `default_x86_64`.
- `roc_serialize` provides helpers for serializing and deserializing to/from bytes.

> - `roc_solve` see [Ambient Lambda Set Specialization](./compiler/solve/docs/ambient_lambda_set_specialization.md) for more information on how polymorphic lambda sets are specialized and resolved in the compiler's type solver. 

Reviewer: ```suggestion
- `roc_solve` The entry point of Roc's type inference system. Implements type inference and specialization of abilities.
```

- `roc_solve_problem` provides types to describe problems that can occur during solving.
- `roc_str` provides `Roc` styled collection reference counting. See [README.md](./compiler/str/README.md) for more information.
- `test_derive` Tests Roc's auto-derivers.
- `test_gen` contains all of Roc's code generation tests. See [README.md](./compiler/test_gen/README.md) for more information.
- `test_mono` Tests Roc's generation of the mono intermediate representation.

> - `test_mono_macros` Macros for use in test_mono

Reviewer: ```suggestion
- `test_mono_macros` Macros for use in test_mono.
```

- `roc_types` Various representations and utilities for dealing with types in the Roc compiler.
- `roc_unify` Implements Roc's unification algorithm, the heartstone of Roc's type inference.

## `docs/` - `roc_docs`

> Generates html documentation from Roc files.
> Used for [roc-lang.org/builtins/Num](https://www.roc-lang.org/builtins/Num).

Reviewer: ```suggestion
Generates HTML documentation for Roc modules.
Used by the [online Roc builtin modules documentation](https://www.roc-lang.org/builtins/Num), and can be used for your Roc code's documentation!
```

## `docs_cli/` - `roc_docs_cli` library and `roc-docs` binary

> Provides a binary that is only used for static build servers.

Reviewer: nit

```suggestion
Provides a binary for building Roc docs, typically only used in Roc's CI. To generate documentation locally, use `roc docs`.
```

## `editor/` - `roc_editor`

> For editing Roc files (Work In Progress). See [README.md](./editor/README.md) for more information.

Reviewer: ```suggestion
Roc's editor. See [README.md](./editor/README.md) for more information.
```

## `error_macros/` - `roc_error_macros`

Provides macros for consistent reporting of errors in Roc's rust code.

## `glue/` - `roc_glue`

> The `roc_glue` crate generates rust code to connect a Roc app with a rust platform. This tool is not necessary for writing a platform in another language, however, it's an added convenience for rust platform devs.

Reviewer: Generally glue will be more generic than just for Rust

```suggestion
The `roc_glue` crate generates code needed for platform hosts to communicate with Roc apps. This tool is not necessary for writing a platform in another language, however, it's a great convenience! Currently supports Rust platforms, and the plan is to support any language via a plugin model.
```

## `highlight/` - `roc_highlight`

Provides syntax highlighting for the editor by transforming a string to markup nodes.

## `linker/` - `roc_linker`

Surgical linker that links platforms to Roc applications.

> We created our own linker for speed and because regular linkers add a lot of problems and complexity. Because we want roc to manage the build system and final linking of the executable, it is significantly less practical to use a regular linker. This can be seen in how many arbitrary libraries we link with the legacy linker because we have no idea what libraries the application is actually using.

Reviewer: ```suggestion
We created our own linker for performance, since regular linkers add complexity that is not needed for linking Roc apps. Because we want `roc` to manage the build system and final linking of the executable, it is significantly less practical to use a regular linker.
```
